scipy/sparse/_compressed.py

```diff
@@ -866,23 +854,58 @@ def _setdiag(self, values, k):
                 max_index = min(M + k, N)
             else:
                 max_index = min(M + k, N, len(values))
-            i = np.arange(max_index, dtype=self.indices.dtype)
+            i = np.arange(-k, max_index - k, dtype=self.indices.dtype)
             j = np.arange(max_index, dtype=self.indices.dtype)
-            i -= k
 
         else:
             if broadcast:
                 max_index = min(M, N - k)
             else:
                 max_index = min(M, N - k, len(values))
             i = np.arange(max_index, dtype=self.indices.dtype)
-            j = np.arange(max_index, dtype=self.indices.dtype)
-            j += k
+            j = np.arange(k, k + max_index, dtype=self.indices.dtype)
 
         if not broadcast:
             values = values[:len(i)]
 
-        self[i, j] = values
+        x = np.atleast_1d(np.asarray(values, dtype=self.dtype)).ravel()
+        if x.squeeze().shape != i.squeeze().shape:
+            x = np.broadcast_to(x, i.shape)
+        if x.size == 0:
+            return
+
+        M, N = self._swap((M, N))
+        i, j = self._swap((i, j))
+        n_samples = x.size
+        offsets = np.empty(n_samples, dtype=self.indices.dtype)
+        ret = csr_sample_offsets(M, N, self.indptr, self.indices, n_samples,
+                                 i, j, offsets)
+        if ret == 1:
+            # rinse and repeat
+            self.sum_duplicates()
+            csr_sample_offsets(M, N, self.indptr, self.indices, n_samples,
+                               i, j, offsets)
+        if -1 not in offsets:
+            # only affects existing non-zero cells
+            self.data[offsets] = x
+            return
+
+        mask = (offsets <= -1)
+        # Boundary between csc and convert to coo
+        if mask.sum() < self.nnz * 0.001:
```

Comment:
Could you include a link to this PR's analysis here? That will be really helpful when later readers wonder why this particular constant was chosen.

Reply:
Good idea -- Done